Add detection for 64bit long int to support padding

richael02 opened the pull request:
Zlib-ng introduces a 32bit padding member in its data
struct to make zlib-ng compatible with zlib, without
which there will be runtime errors when making zlib-ng
a binary replacement of zlib. Now the padding is enabled
only when both ZLIB_COMPAT and X86_64 are defined, so it
needs support for other platforms where sizeof(long int)
is not equal to sizeof(int).

This patch fixes the problem by adding the detection for
64bit long int.

Signed-off-by: Richael Zhuang <richael.zhuang@arm.com>

richael02 committed:
Add detection for 64bit long int to support padding

Zlib-ng introduces a 32bit padding member in its data
struct to make zlib-ng compatible with zlib, without
which there will be runtime errors when making zlib-ng
a binary replacement of zlib. Now the padding is enabled
only when both ZLIB_COMPAT and X86_64 are defined, so it
needs support for other platforms where sizeof(long int)
is not equal to sizeof(int).

This patch fixes the problem by adding the detection for
64bit long int.

Signed-off-by: Richael Zhuang <richael.zhuang@arm.com>

mtl1979 commented on CMakeLists.txt:
```diff
@@ -368,7 +385,7 @@ set(ZLIB_ARCH_SRCS)
 if("${ARCH}" MATCHES "x86_64" OR "${ARCH}" MATCHES "AMD64")
     set(ARCHDIR "arch/x86")
-    add_definitions(-DX86_64 -DX86_NOCHECK_SSE2 -DUNALIGNED_OK -DUNROLL_LESS)
+    add_definitions(-DX86_NOCHECK_SSE2 -DUNALIGNED_OK -DUNROLL_LESS)
```
This will break the assembler code.

mtl1979 commented on configure:
```diff
@@ -769,8 +790,8 @@ case "${ARCH}" in
-                CFLAGS="${CFLAGS} -DX86_64 -DX86_NOCHECK_SSE2"
-                SFLAGS="${SFLAGS} -DX86_64 -DX86_NOCHECK_SSE2"
+                CFLAGS="${CFLAGS} -DX86_NOCHECK_SSE2"
+                SFLAGS="${SFLAGS} -DX86_NOCHECK_SSE2"
```
Bad idea... See above.

richael02 replied:
the flag "X86_64" is only covered in these places:

CMakeLists.txt:383:    add_definitions(-DX86_64 -DX86_NOCHECK_SSE2 -DUNALIGNED_OK -DUNROLL_LESS)
configure:772:                CFLAGS="${CFLAGS} -DX86_64 -DX86_NOCHECK_SSE2"
configure:773:                SFLAGS="${SFLAGS} -DX86_64 -DX86_NOCHECK_SSE2"
zlib.h:112:#if defined(ZLIB_COMPAT) && defined(X86_64)

So from my view it's only defined for enabling padding

mtl1979 replied:
The define has existed long time before the padding was added... Patch for one architecture shouldn't touch code for another architecture. That however doesn't include lines that apply for multiple architectures and only change is adding more tests.

mtl1979 commented on zlib.h:
```diff
@@ -109,8 +109,9 @@ typedef struct z_stream_s {
-    uint32_t              padding;    /* pad out to the same size as the zlib struct */
+#if defined(ZLIB_COMPAT) && defined(IS_64BIT_LONG)
+    uint32_t              padding;    /* pad out to the same size as the zlib struct: padding
+                                         is needed when the size of long int is 64 bits */
```
This code was incorrect originally and will be removed very soon. Assumes little-endian order which isn't true for all supported processors.

mtl1979 reviewed (changes requested):
Conflicts with changes in other pull requests. Breaks compiling on AMD64 platform.

richael02 committed:
Add padding support for Arm architectures

zlig-ng introduces a 32 bit padding member in its data
struct to make zlib-ng compatible with zlib, without
which there will be runtime errors when making zlib-ng
a binary replacement of zlib.The padding is only enabled
when ZLIB_COMPAT and X86_64 are defined, so it needs
support for Arm architectures.

This patch fixes the problem by adding the detection for
64bit long int.

Signed-off-by: Richael Zhuang <richael.zhuang@arm.com>

richael02 commented:
So this new commit only adds the support our ARM platforms .And I have verified it.

mtl1979 commented on zlib.h:
```diff
@@ -109,8 +109,9 @@ typedef struct z_stream_s {
-    uint32_t              padding;    /* pad out to the same size as the zlib struct */
+#if defined(ZLIB_COMPAT) && (defined(X86_64) || defined(ARM_64BIT_LONG))
+    uint32_t              padding;    /* pad out to the same size as the zlib struct: padding
+                                         is needed when the size of long int is 64 bits */
```
This code is still wrong and pending removal within a week... The problem is that compiler automatically adds extra padding after variable "padding" even though the extra padding should be between "data_type" and "adler" on little-endian machines because both are 32-bit variables in zlib-ng, but "adler32" is 64-bit in zlib so still aligned differently.

richael02 replied:
I don't know how will you deal with padding finally . What I do is just making zlib-ng work well on out platforms. So I will wait for your final modification and then look into it.

mtl1979 replied:
ARM is little-endian by default, so you should have padding before and after "adler32"... both should be uint32_t when long int is 64 bits.
We will use two versions for zlib.h and the one used for ZLIB_COMPAT will be reverted to use same types as original zlib.

Dead2 commented:
zlib.h / zlib-ng.h changes needs new review

mtl1979 commented:
Unnecessary with latest merges.
